File: Week_02/G20200343030632/Leetcode_589_632.py
#
# @lc app=leetcode.cn id=589 lang=python3
#
# [589] N叉树的前序遍历
#

# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        return self.preorder_v2(root)
    def preorder_v1(self, root: 'Node') -> List[int]:
        res = []
        def helper(root: 'Node', res: List[int]):
            if root:
                res.append(root.val)
                for child in root.children:
                    helper(child, res)
        helper(root, res)
        return res
    def preorder_v2(self, root):
        res, stack = [], root and [root]
        while stack:
            cur = stack.pop()
            res.append(cur.val)
            stack += [child for child in cur.children[::-1] if child]
        return res
        
# @lc code=end

File: Week_02/G20200343030632/Leetcode_589_632.java
import java.util.Deque;
import java.util.LinkedList;

/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    public List<Integer> preorder(Node root) {
        return preorder_v2(root);
    }
    public List<Integer> preorder_v1(Node root) {
        List<Integer> res = new LinkedList<>();
        helper(root, res);
        return res;
    }
    private void helper(Node node, List<Integer> res) {
        if (node != null) {
            res.add(node.val);
            for (Node n : node.children) {
                helper(n, res);
            }
        }
    }
    public List<Integer> preorder_v2(Node root) {
        // 迭代
        List<Integer> res = new LinkedList<>();
        if (root == null) {
            return res;
        }
        Deque<Node> stack = new LinkedList<>();
        stack.addFirst(root);
        while (!stack.isEmpty()) {
            Node node = stack.removeFirst();
            res.add(node.val);
            for (int i = node.children.size() - 1; i >= 0; i--) {
                stack.addFirst(node.children.get(i));
            }

        }
        return res;
    }

    
}